dataplane/protocol/lldp/lldp.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lldp
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/openconfig/ygnmi/ygnmi"
+
+	"github.com/openconfig/lemming/gnmi/gnmiclient"
+	"github.com/openconfig/lemming/gnmi/oc"
+
+	log "github.com/golang/glog"
+
+	"github.com/openconfig/lemming/gnmi/oc/ocpath"
+
+	"github.com/openconfig/lemming/dataplane/proto/packetio"
+)
+
+// Daemon is the implementation of the LLDP protocol.
+type Daemon struct {
+	enabled     bool                   // whether LLDP is enabled globally
+	portEnabled map[string]bool        // contains the enabled ports
+	portDaemons map[string]*portDaemon // tracks the active port daemons
+}
+
+// New returns the main procotol handler.
+func New() *Daemon {
+	return &Daemon{
+		portEnabled: map[string]bool{},
+		portDaemons: map[string]*portDaemon{},
+	}
+}
+
+// Start starts the procotol handler.
+func (d *Daemon) Start() {
+	if d.portDaemons == nil {
+		d.portDaemons = map[string]*portDaemon{}
+	}
+	d.enabled = true
+}
+
+// Stop stops the procotol handler by stopping all port daemons.
+func (d *Daemon) Stop() {
+	for _, p := range d.portDaemons {
+		p.Stop()
+	}
+	d.enabled = false
+	d.portDaemons = nil
+}
+
+// changePortState tries to change the port state and returns whether the state is acutally changed.
+func (d *Daemon) changePortState(intf *oc.Lldp_Interface) bool {
+	// Update the PD state.
+	wantActive := d.enabled && intf.GetEnabled()
+	pd, currActive := d.portDaemons[intf.GetName()]
+	switch {
+	case !currActive && wantActive:
+		d.portDaemons[intf.GetName()] = newPortDaemon(intf.GetName())
+	case currActive && !wantActive:
+		pd.Stop()
+		delete(d.portDaemons, intf.GetName())
+	}
+
+	// Update the enabled state.
+	state, ok := d.portEnabled[intf.GetName()]
+	if !ok {
+		d.portEnabled[intf.GetName()] = false
+	}
+	if state != intf.GetEnabled() {
+		d.portEnabled[intf.GetName()] = intf.GetEnabled()
+		return true
+	}
+	return false
+}
+
+func (d *Daemon) reconcileLldpEnabled(sb *ygnmi.SetBatch, intent *oc.Root, c *ygnmi.Client) error {
```

Comment:
nit: these func could be folded into the Reconcile() func, they are pretty simple

Reply:
Done